### PR TITLE
More Improvements to Boxplot dialog

### LINKED ISSUE #8759: Yet more corrections and improvements in boxplot dialog
There are important improvements and corrections, to each of the buttons, namely Boxplot, Jitter, Violin.  Here is the current boxplot button:

image

and here is the violin plot button:

image

and finally, here is Jitter:

image

Now the improvements:
a) There is a Swap Parameters checkbox, disabled, for each option.  We don't need it.  It should be deleted.
b) For the boxplot, move the Varible Width Checkbox to where the Swap Parameters is now.
c) This next change is identical for boxplot and violin plot.  Ideally leave jitter unchanged.  (If that is tricky then it isn't a big problem if the first parts are for all 3.
   1) Change the factor receiver so it can also receive any type of variables.
   2) Ideally change the label to `Factor (Usually):`
   3) Add a checkbox that is disabled for non-numeric or date. It is checked by default. It has a label Width followed by a numeric field into which any number can be typed.  It is not an up-down.  Perhaps have 0.25 as a default.  (This assumes it won't be implemented if disabled.)
   4) If enabled and checked, then it adds `aes(group = cut_width(carat, 0.25))` into the command generated.  This could be included in the aesthetics or given as an aditional aesthetic in the geom_boxplot().  It is taken from [here](https://ggplot2.tidyverse.org/reference/geom_boxplot.html).  Change carat and 0,25 for the values given in the dialog.
d) Just for boxplot now, there is an option to `Add Points`.  This adds the jittered points.  It is a bit odd that this means the outliers are given twice.  There is now a nice option for the boxplots, namely `geom_boxplot(outlier_shape=NA)`.  Add this when the option is ticked!  This is all great!
e) Finally for all three buttons the code is incorrect for the groups to connect option if the second factor is included.  It gives just a single line, and should give 3 lines, in the example below:

image

I will insert the new code later here.

f) In the Violin  and the Jitter options, add a checkbox above the Add Points option with label Add Boxplot.  Default Unchecked. If checked it shows a label Width, follwed by an Up-down control.  This goes from 0 to 1 in steps of 0.01.  Default is 0.5.  It adds ggplot2::geom_boxplot(width = 0.5) to the code generated.

### PR DESCRIPTION
**Pull request by Vitalis95**
Fixes #8759 
@rdstern , @N-thony - Added  `Width, Add Boxplot` controls,  removed one of the outliers that are given twice in Boxplot

### CODE REVIEW

**Review by rdstern — COMMENTED** [review]
@Vitalis95 looks very good indeed.
a) You have `outlier_shape=NA`, which doesn't hide the outlier.  Change to `outlier.shape=NA`. (Sorry that was my mistake in the issue.)
b)  And I didn't add the code for the groups when there are 2 factors.  Here is the [reference that gives the solution](https://stackoverflow.com/questions/57043433/how-to-connect-the-median-values-of-a-boxplot-with-multiple-groups-using-lines-i).  I hope you can puzzle it out from here? Tell me if not.

**Comment by Vitalis95** [comment]
> @Vitalis95 looks very good indeed. (In fact all seems to be working well, except the bits I got wrong or omitted!) a) You have `outlier_shape=NA`, which doesn't hide the outlier. Change to `outlier.shape=NA`. (Sorry that was my mistake in the issue.) b) And I didn't add the code for the groups when there are 2 factors. Here is the [reference that gives the solution](https://stackoverflow.com/questions/57043433/how-to-connect-the-median-values-of-a-boxplot-with-multiple-groups-using-lines-i). I hope you can puzzle it out from here? Tell me if not.

@rdstern , Can we have a call concerning **b)**?

**Comment by rdstern** [comment]
ok @Vitalis95 here is an example:
image

This is the code we currently produce:

```
# Dialog: Boxplot

survey <- data_book$get_data_frame(data_name="survey")
last_graph <- ggplot2::ggplot(data=survey, mapping=ggplot2::aes(y=yield, x=variety, fill=village)) + 
ggplot2::geom_boxplot(outlier.colour="red") + ggplot2::stat_summary(geom="line", group=1, fun.y="median", size=1.5, color="blue") + theme_grey()
data_book$add_object(data_name="survey", object_name="last_graph", object_type_label="graph", object_format="image", object=check_graph(graph_object=last_graph))
data_book$get_object_data(data_name="survey", object_name="last_graph", as_file=TRUE)
rm(list=c("last_graph", "survey"))
```

The stat_summary needs to change.  Here is the solution for that part of the code:

`ggplot2::stat_summary(geom="line", aes(group=village,colour=village), fun.y="median", size=1.5,position = position_dodge(width = 0.9))`

The colour and group now go into the aes, and you add the position dodge as well.

The result I get is here:

image

**Comment by Vitalis95** [comment]
@rdstern ,I add aes if `factor `and `second factor`  receivers are not empty? ` aes(group=village,colour=village)`
What if we only variable in` factor` or `second factor` ?

**Comment by rdstern** [comment]
@Vitalis95 I don't see your ewcwent message here, but it is an excellent observation that I had forgotten.  We sometimes have the same variable in the x and fill, for example Villagre in both.  In that case it is just a single factor and the exisating code works fine.  Iy is only when they are different variables that the changes are needed.

**Comment by Vitalis95** [comment]
@rdstern , have a look at it

**Review by rdstern — DISMISSED** [review]
@Vitalis95 very well done.  @N-thony can this be merged?

File: instat/dlgBoxPlot.vb [hunk 1]
```diff
@@ -128,8 +136,8 @@ Public Class dlgBoxplot
 
         ucrByFactorsReceiver.SetParameter(New RParameter("x", 1))
         ucrByFactorsReceiver.Selector = ucrSelectorBoxPlot
-        ucrByFactorsReceiver.SetIncludedDataTypes({"factor"})
-        ucrByFactorsReceiver.strSelectorHeading = "Factors"
+        'ucrByFactorsReceiver.SetIncludedDataTypes({"factor"})
+        'ucrByFactorsReceiver.strSelectorHeading = "Factors"
```
[the remark this answers is not in the document]
Comment (Vitalis95) [reply instat/dlgBoxPlot.vb]: Done

**Review by rdstern — APPROVED** [review]
@N-thony I am approving again

**Pull request merged**